### PR TITLE
Enable latest-Recommended analysis rules

### PR DESCRIPTION
**Pull request by lahma**
* move Directory.Build.props to root
* fix some issues

### CODE REVIEW

File: src/Parlot/Fluent/ParseContext.cs [hunk 1]
```diff
@@ -4,7 +4,9 @@ namespace Parlot.Fluent
 {
     public class ParseContext
     {
-        public static int DefaultCompilationThreshold = 0;
+#pragma warning disable CA2211 // Non-constant fields should not be visible
+        public static int DefaultCompilationThreshold;
```
Comment (sebastienros) [inline src/Parlot/Fluent/ParseContext.cs:8]: Need to check that. Maybe the `0` is something I forgot to change. And also I don't see why this couldn't be `const` (without checking the rest of the code).

File: src/Parlot/Compilation/ExpressionHelper.cs [hunk 2]
```diff
@@ -37,7 +37,9 @@ public static class ExpressionHelper
         public static Expression ArrayEmpty<T>() => ((Expression<Func<object>>)(() => Array.Empty<T>())).Body;
         public static Expression New<T>() where T : new() => ((Expression<Func<T>>)(() => new T())).Body;
         
+#pragma warning disable CA2211 // Non-constant fields should not be visible
         public static Expression<Func<Cursor, char, char, bool>> CharacterIsInRange = (cursor,b,c) => Character.IsInRange(cursor.Current, b, c);
```
Comment (sebastienros) [inline src/Parlot/Compilation/ExpressionHelper.cs:41]: This could be a readonly field, like all these entries.

File: Directory.Build.props [hunk 3]
```diff
@@ -21,4 +21,22 @@
...
 
+  <PropertyGroup Label="Analysis rules">
+
+    <AnalysisLevel>latest-Recommended</AnalysisLevel>
+
+    <!-- Prefix generic type parameter name with 'T'  -->
+    <NoWarn>$(NoWarn);CA1715;</NoWarn>
+
+    <!-- Do not declare visible instance fields  -->
+    <NoWarn>$(NoWarn);CA1051;</NoWarn>
```
Comment (sebastienros) [inline Directory.Build.props:32]: If this is for the CompilationHelper then maybe I should just use properties or readonly fields.

Comment (lahma) [reply Directory.Build.props:32]: It's also all structs which don't use properties but public fields.

File: src/Parlot/Fluent/TextLiteral.cs [hunk 4]
```diff
@@ -40,7 +42,7 @@ public TextLiteral(string text, StringComparison comparisonType)
                 }
                 else
                 {
-                    ExpectedChars = ignoreCase ? [Text.ToUpper()[0], Text.ToLower()[0]] : [Text[0]];
+                    ExpectedChars = ignoreCase ? [Text.ToUpper(CultureInfo.CurrentCulture)[0], Text.ToLower(CultureInfo.CurrentCulture)[0]] : [Text[0]];
```
Comment (sebastienros) [inline src/Parlot/Fluent/TextLiteral.cs:45]: When you read the code this makes totally sense. At first I thought this was evil but this is exactly what the argument means. Good job.

**Review by sebastienros — APPROVED** [review]

**Pull request merged**